soroban-env-host/src/test/invocation.rs

```diff
@@ -29,6 +29,34 @@ fn invoke_single_contract_function() -> Result<(), HostError> {
     Ok(())
 }
 
+#[test]
+fn invoke_alloc() -> Result<(), HostError> {
+    let host = Host::test_host_with_recording_footprint();
+    host.enable_debug()?;
+    let contract_id_obj = host.register_test_contract_wasm(ALLOC);
+    let res = host.call(
+        contract_id_obj,
+        Symbol::try_from_small_str("sum")?,
+        host.test_vec_obj::<u32>(&[128])?,
+    )?;
+    assert!(res.shallow_eq(&8128_u32.into()));
+    let used_bytes = host.budget_cloned().get_mem_bytes_consumed()?;
+    // The general pattern of memory growth in this contract will be a sequence
+    // of vector-doublings, but these are masked by the fact that we only see
+    // the calls that cause the backing vector of wasm linear memory to grow,
+    // which happens as the guest vector crosses 64k boundaries (and eventually
+    // starts growing in steps larger than 64k itself).
+    //
+    // So we wind up with a growth-sequence that's a bit irregular: +0x10000,
+    // +0x20000, +0x30000, +0x50000, +0x90000. Total is 1 + 2 + 3 + 5 + 9 = 20
+    // pages or about 1.3 MiB, plus the initial 17 pages (1.1MiB) plus some more
+    // slop from general host machinery allocations, we get around 2.5MiB. Call
+    // is "less than 3MiB".
+    assert!(used_bytes > (128 * 4096));
+    assert!(used_bytes < 0x30_0000);
```

Comment:
Where does the initial 17 pages come from? Is that due to our allocator implementation?

Reply:
I don't mind being a little loose in the estimate here since the toolchain can change its mind about where it starts the heap and rust's `Vec` implementation could change its mind about its growth policy and so forth.

The initial 17 pages is the default linear memory map given to a Rust program compiled to wasm. Its content is [described here](https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/spec/wasm_base.rs#L8-L28) but briefly: there's an initial blank page (I guess this is just the common thing of making null pointers point to a decently-large region of zeroes) followed by static data followed by 1MiB of stack (which grows down). Since there's no static data in this example, we get "just" a 1MiB stack (16 64k pages) after the initial blank page, or 17 pages total.